core/src/main/java/org/apache/iceberg/BaseChangelogContentScanTask.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.expressions.ResidualEvaluator;
+import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
+
+abstract class BaseChangelogContentScanTask<ThisT, F extends ContentFile<F>>
+    extends BaseContentScanTask<ThisT, F>
+    implements ChangelogScanTask {
+
+  private final int changeOrdinal;
+  private final long commitSnapshotId;
+
+  BaseChangelogContentScanTask(int changeOrdinal, long commitSnapshotId, F file,
+                               String schemaString, String specString, ResidualEvaluator residuals) {
+    super(file, schemaString, specString, residuals);
+    this.changeOrdinal = changeOrdinal;
+    this.commitSnapshotId = commitSnapshotId;
+  }
+
+  @Override
+  public int changeOrdinal() {
+    return changeOrdinal;
+  }
+
+  @Override
+  public long commitSnapshotId() {
+    return commitSnapshotId;
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("change_ordinal", changeOrdinal)
+        .add("commit_snapshot_id", commitSnapshotId)
+        .add("file", file().path())
+        .add("partition_data", file().partition())
+        .add("residual", residual())
+        .toString();
+  }
+
+  abstract static class SplitScanTask<ThisT, T extends ContentScanTask<F> & ChangelogScanTask, F extends ContentFile<F>>
```

Comment:
@szehon-ho, instead of making the base class abstract and extending it everywhere, I can probably accept a closure to construct a split scan task in the base class. Then we won't need any subclasses. I think that's going to be cleaner. What do you think?

Reply:
That sounds great

Reply:
I managed to implement it. Could you take a look, @szehon-ho?